drake/solvers/NloptSolver.cpp

```diff
@@ -0,0 +1,269 @@
+          // thing to do if a decision variable is bound to multiple
+          // objectives?
+          grad[j] += ty(0).derivatives()(j);
+        }
+      }
+    }
+  }
+
+  return cost;
+}
+
+/// Structure to marshall data into the NLopt callback functions,
+/// which take only a single pointer argument.
+struct WrappedConstraint {
+  const Constraint* constraint;
+  const VariableList* variable_list;
+};
+
+double EvaluateScalarConstraint(const std::vector<double>& x,
+                                std::vector<double>& grad,
+                                void* f_data) {
+  const WrappedConstraint* wrapped =
+      reinterpret_cast<WrappedConstraint*>(f_data);
+
+  Eigen::VectorXd xvec = MakeEigenVector(x);
+
+  const Constraint* c = wrapped->constraint;
+  const size_t num_constraints = c->getNumConstraints();
+  assert(num_constraints == 1);
+
+  TaylorVecXd ty(1);
+  TaylorVecXd this_x = MakeInputTaylorVec(xvec, *(wrapped->variable_list));
+  c->eval(this_x, ty);
+  double result = ty(0).value() * -1;
+
+  const Eigen::VectorXd& upper_bound = c->getUpperBound();
+  if (upper_bound[0] != std::numeric_limits<double>::infinity()) {
+    result -= upper_bound[0];
+  }
+
+  // TODO sam.creasey How should lower bounds be handled?  We don't
+  // have any tests which use them...
+
+  if (!grad.empty()) {
+    grad.assign(grad.size(), 0);
+    for (const DecisionVariableView& v : *(wrapped->variable_list)) {
+      for (size_t j = v.index(); j < v.index() + v.size(); j++) {
+        grad[j] = ty(0).derivatives()(j) * -1;
+      }
+    }
+  }
+  return result;
+}
+
+void EvaluateVectorConstraint(unsigned m, double* result, unsigned n,
+                              const double* x, double* grad, void* f_data) {
+  const WrappedConstraint* wrapped =
+      reinterpret_cast<WrappedConstraint*>(f_data);
+
+  Eigen::VectorXd xvec(n);
+  for (size_t i = 0; i < n; i++) {
+    xvec[i] = x[i];
+    if (grad) { grad[i] = 0; }
+  }
+
+  const Constraint* c = wrapped->constraint;
+  const size_t num_constraints = c->getNumConstraints();
+  assert(num_constraints == m);
+
+  TaylorVecXd ty(m);
+  TaylorVecXd this_x = MakeInputTaylorVec(xvec, *(wrapped->variable_list));
+  c->eval(this_x, ty);
+
+  const Eigen::VectorXd upper_bound = c->getUpperBound();
+  for (size_t i = 0; i < num_constraints; i++) {
+    result[i] = ty(i).value() * -1;
+    if (upper_bound[i] != std::numeric_limits<double>::infinity()) {
+       result[i] -= upper_bound[i];
+     }
+  }
+
+  // TODO sam.creasey How should lower bounds be handled?  We don't
+  // have any tests which use them...
+
+  if (grad) {
+    for (const DecisionVariableView& v : *(wrapped->variable_list)) {
+      for (size_t i = 0; i < num_constraints; i++) {
+         for (size_t j = v.index(); j < v.index() + v.size(); j++) {
+           // TODO sam.creasey I don't actually know if summing these
+           // is the right thing to do when more than one constraint
+           // is present.
```

Comment:
This seems to be the same as the case for the gradients in the objective function (handled above).